ghcide/src/Development/IDE/Core/Rules.hs

```diff
@@ -689,6 +723,17 @@ ghcSessionDepsDefinition file = do
         res <- liftIO $ newHscEnvEqWithImportPaths (envImportPaths env) session' []
         return ([], Just res)
 
+
+-- This function is also responsible for ensuring database consistency
+-- Whenever we read a `.hi` file, we must check to ensure we have also
+-- indexed the corresponding `.hie` file. If this is not the case (for example,
+-- `ghcide` could be killed before indexing finishes), we must re-index the
+-- `.hie` file. Most of the time, there should be an up2date `.hi` file on
+-- disk since we are careful to write out the `.hie` file before writing the
+-- `.hi` file
+-- If we don't have a `.hi` file, the `regenerateHiFile` function is responsible
+-- for generating both a fresh `hi` and `hie` file, and queueing up a
+-- index operation for the `.hie` file.
 getModIfaceFromDiskRule :: Rules ()
 getModIfaceFromDiskRule = defineEarlyCutoff $ \GetModIfaceFromDisk f -> do
```

Comment:
The `GetModIfaceFromDisk` rule has too many responsibilities. 
This makes it difficult to replace with e.g. a custom rule that delegates `.hi` and `.hie` files to a cloud build system. 
Could it be broken up into two or more rules with well defined responsibilities ?

Reply:
I don't think another rule makes sense. I could factor out the database consistency checking logic out into another function, not sure if that would help.

Reply:
Would it be possible to extract all the indexing logic into a new rule `IndexedModule` that depends on `GetModIfaceFromDisk`? It is a rule and not just a function so that Shake ensures consistency and avoids duplicate work.

Then you can either make the dependency on `IndexedModule` explicit, or you can create a new rule `GetModIfaceFromDiskAndIndex` that does both. 

Can you see any reason not to do that?

Reply:
I don't really see the benefit. What will depend on this rule?

I prefer the logic to be tied to `GetModIfaceFromDisk`. This way, we can ensure the index is always up to date.

Given a file `f`, either

1. `f` is a FOI, so we don't need to index it in the database
2. It is not a FOI (this case calls into `GetModIfaceFromDisk` when the file is attempted to be loaded), but
    a. We don't have a `.hi`/`.hie` file - we generate and index the `.hie` file too (the only function to write a `.hie` file is `writeAndIndexHieFile`), so the database consistency is maintained)
    b. we have a `.hi` file - In this case we need to check that the corresponding `.hie` file has been indexed - this is what all the extra logic in `GetModIfaceFromDisk` is.

My concern with splitting this up into rules is that this is decoupled, so we might have files being loaded/compiled, without ensuring the database is notified of these changes. 

I also don't see what duplicate work we could be avoiding by doing this.

Additionally, in 2(b), if the `.hi` file exists but the `.hie` file doesn't for some reason (this shouldn't happen, but could if for example someone switched from an old version of ghcide to a newer one), we regenerate both the `.hi` and `.hie` files. This modifies the result of the `GetModIfaceFromDisk` rule, so if this logic is living in a different rule that depends on `GetModIfaceFromDisk` then the iface files returned by `GetModIfaceFromDisk` and `IndexedModule` might diverge.

Reply:
The benefit of having simple rules with clear responsibilities is that an application can override the `GetModIfaceFromDisk` to generate the interface file by other means, like an external build system with a cloud cache. This is a real use case that arises in our work codebase.

Why do you have concerns that splitting will lead to inconsistency? Based on the experience of splitting`GetModSummary` and `GetModSummaryWithoutTimestamps`, this should not be an issue. Moreover, `GetModIfaceFromDisk` is only used inside the `Rules` module, there are no external callers.

I don't understand your last point: 
>  This modifies the result of the GetModIfaceFromDisk rule, so if this logic is living in a different rule that depends on GetModIfaceFromDisk then the iface files returned by GetModIfaceFromDisk and IndexedModule might diverge.

Why would `IndexedModule` return ani face file?

Reply:
What would use `IndexedModule`?

> Why would IndexedModule return an iface file?

What happens when `IndexedModule` discovers that it needs to regenerate both the `.hi` and `.hie` file?

> The benefit of having simple rules with clear responsibilities is that an application can override the GetModIfaceFromDisk rule definition to generate the interface file by other means, like an external build system with a cloud cache

I still think its better if whoever overrides `GetModIfaceFromDisk` also ensures that the indexing operations for the database are queued up.

Reply:
As it stands now, `GetModIfaceFromDisk` doesn't ensure that the module is indexed, it only ensures that either the module is indexed, or we have queued an operation for indexing the module.

Would `IndexModule` actually have to wait to ensure the module is indexed?

Reply:
> What would use `IndexedModule`?

Any rule that requires up-to-date-ness of the hiedb. It doesn't matter whether `IndexedModule` returns `()` or some actual thing like a database id.

> 
> What happens when `IndexedModule` discovers that it needs to regenerate both the `.hi` and `.hie` file?
> 

It's an error, so throw an exception?
The way I see it, `IndexedModule` cannot ever need to regenerate the `.hi` file, only the `.hie` file, it's just that both get generated at the same time. If `IndexedModule` depends on `GetModIfaceFromDisk`, then by the contract there must be an up-to-date `.hie` file on disk, and if there isn't then it's an error and you can throw, right?

> As it stands now, GetModIfaceFromDisk doesn't ensure that the module is indexed, it only ensures that either the module is indexed, or we have queued an operation for indexing the module. Would IndexModule actually have to wait to ensure the module is indexed?

I'm not asking for a change of semantics. It can keep the same semantics as it has now.

Reply:
> Any rule that requires up-to-date-ness of the hiedb. It doesn't matter whether IndexedModule returns () or some actual thing like a database id.

That would be the hover/definition/etc handlers. which use `useE` and `useWithStale`. Because of this, I don't see what purpose `useE IndexedModule` would serve there - either we block waiting for the index to be completed (bad, and negates all the advantages of the persistent rule), or we don't, and we might as well not have called the rule.

The other option is to have it called from `GetModIface` right after it calls `GetModIfaceFromDisk`. In this case, I don't see what benefit the rule gives us over just factoring out the logic into a separate function.

Reply:
Calling a non blocking variant of `useE IndexedModule` from hover and getDefinition handlers serves a purpose: it makes the dependency explicit. The rule returns nothing because the database has mutable state, but it could return a version number, or a handle to a database partition, etc. 

Extracting to a helper function seems less elegant but perhaps more efficient. Just keep in mind that much of this stuff is internal and you may have to expose some of the hiedb machinery.

Reply:
If the rule is called on requests, the database will almost always be out of date - indexing will be queued up on the request, then the request will be immediately responded to using stale information (since we don't block waiting for the index to complete), then there will be more edits that make the database out of date again until there is another request.

Compare this to triggering the index on edit (when `GetModIface`/`GetModIfaceFromDisk` is requested by `kick`). Then, there is a good chance that the database is up to date on any given request.

Reply:
You can always call the rule in `kick`

Reply:
I'm implementing this right now and realized another consequence - we won't respect the `CheckProject` settings anymore, since we need to call it for all known files (not just FOIs), which would transitively cause `GetModIface` to be called on all known files every kick.

The more I think about it, the more the "push" model makes sense to me, where we update the database whenever we realize there are some changes, rather than the "pull" model where we periodically poll all dependencies to check if anything has changed and then update the database.

Reply:
That makes sense, but you can define a new rule `GetModIfaceFromDiskAndIndex` and use it in all the call sites of `GetModIfaceFromDisk`. To be clear, there's no need to have two new rules `IndexedModule` and `GetModIfaceFromDiskAndIndex`, just the latter.

Reply:
I've added `GetModIfaceFromDiskAndIndex`, but still not entirely convinced why it needs to be a rule.

Reply:
Awesome, thanks! It looks good to me :)